### PR TITLE
feat!: Refactor ownership pallet to transfer ownership of spaces, posts and domains

### PR DESCRIPTION
**Pull request by F3Joule**
### Release Checklist

- [x] Verify `spec_version` has been incremented since the last release.
- [x] Verify pallet and extrinsic ordering has stayed the same. Bump `transaction_version` if not.
- Verify new extrinsics have been correctly whitelisted/blacklisted for `BaseCallFilter`
( #257 )
- [x] There are benchmarks for newly created extrinsics.
- [ ] Verify weights have been updated for any modified runtime logic.
- [x] There are tests for newly created logic.
- [x] All the tests pass.
- [x] Ensure migrations run correctly.

### CODE REVIEW

**Comment by F3Joule** [comment]
Try-runtime results:

```
2024-02-27 13:15:29 initialized state externalities with storage root 0x635bbe399c9eb44fe2052d6bab0667cea4a913971fbd8540945fa701d2766e34 and state_version V0
2024-02-27 13:15:29 original spec: RuntimeString::Owned("subsocial-parachain")-40, code hash: 8a405e57cce8bd1cca762f61406c015023a3612c4b11e96a4e61afca9f5deace
2024-02-27 13:15:29 new spec: RuntimeString::Owned("subsocial-parachain")-41, code hash: ecd313b2566d42c1eb17996075a3c7eeb3eab294f41b1bf0f2f1d05533a4005b
2024-02-27 13:15:29 Onchain storage version: StorageVersion(0), current storage version: StorageVersion(1)
2024-02-27 13:15:29 Running migration with current storage version StorageVersion(1) / onchain StorageVersion(0)
2024-02-27 13:15:29 Upgraded 4 records, storage to version StorageVersion(1)
2024-02-27 13:15:29 ⚠️ ParachainSystem declares internal migrations (which *might* execute). On-chain `StorageVersion(2)` vs current storage version `StorageVersion(2)`
2024-02-27 13:15:29 ⚠️ XcmpQueue declares internal migrations (which *might* execute). On-chain `StorageVersion(3)` vs current storage version `StorageVersion(2)`
2024-02-27 13:15:29 ⚠️ PolkadotXcm declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `StorageVersion(1)`
2024-02-27 13:15:29 ⚠️ DmpQueue declares internal migrations (which *might* execute). On-chain `StorageVersion(2)` vs current storage version `StorageVersion(1)`
2024-02-27 13:15:30 TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = (825000000 ps, 0 byte), total weight = (500000000000 ps, 5242880 byte) (0.17 %, 0.00 %).
```

**Comment by F3Joule** [comment]
Extirnsics ordering (metadata comparison) check results:

```
----------------------------------------------------------------------

Metadata comparison:
Date: Tue 27 Feb 2024 13:23:09 EET
Reference: wss://para.subsocial.network
Target version: subsocial-collator 0.2.2-6020d1c4e23
Chain: dev
----------------------------------------------------------------------
              [Spec] name: subsocial-parachain
                     spec_version: 40 -> 41
                     transaction_version: 8 -> 9
          [Metadata] version: 14
           [Modules] num: 38
                          [System] idx: 0 (calls: 8, storage: 17)
                 [ParachainSystem] idx: 1 (calls: 4, storage: 22)
                       [Timestamp] idx: 3 (calls: 1, storage: 3)
                        [Balances] idx: 10 (calls: 6, storage: 6)
                        [Treasury] idx: 12 (calls: 5, storage: 5)
               [CollatorSelection] idx: 21 (calls: 5, storage: 6)
                         [Session] idx: 22 (calls: 2, storage: 8)
                         [Vesting] idx: 26 (calls: 5, storage: 3)
                           [Proxy] idx: 27 (calls: 10, storage: 3)
                        [Multisig] idx: 29 (calls: 4, storage: 2)
                       [XcmpQueue] idx: 30 (calls: 9, storage: 11)
                     [PolkadotXcm] idx: 31 (calls: 10, storage: 12)
                        [DmpQueue] idx: 33 (calls: 1, storage: 6)
                         [Domains] idx: 60 (calls: 9, storage: 8)
                          [Energy] idx: 61 (calls: 2, storage: 4)
                       [FreeProxy] idx: 62 (calls: 1, storage: 2)
                  [CreatorStaking] idx: 63 (calls: 14, storage: 13)
                    [EvmAddresses] idx: 64 (calls: 2, storage: 3)
             [ResourceDiscussions] idx: 65 (calls: 2, storage: 2)
                           [Roles] idx: 71 (calls: 8, storage: 6)
                  [AccountFollows] idx: 72 (calls: 3, storage: 4)
                        [Profiles] idx: 73 (calls: 3, storage: 2)
                    [SpaceFollows] idx: 74 (calls: 3, storage: 4)
                  [SpaceOwnership] idx: 75 (calls: 3, storage: 2)
                                   [+] calls: transferOwnership
                                   [-] calls: transferSpaceOwnership
                        [acceptPendingOwnership] idx: 1 (args: 1)
                                                 (u64) -> (PalletOwnershipEntityWithOwnership)
                        [rejectPendingOwnership] idx: 2 (args: 1)
                                                 (u64) -> (PalletOwnershipEntityWithOwnership)
                                   [+] storage: pendingOwnershipTransfers
                                   [-] storage: pendingSpaceOwner
                          [Spaces] idx: 76 (calls: 4, storage: 4)
                     [PostFollows] idx: 77 (calls: 2, storage: 4)
                           [Posts] idx: 78 (calls: 6, storage: 6)
                       [Reactions] idx: 79 (calls: 6, storage: 5)
                            [Sudo] idx: 255 (calls: 4, storage: 2)

------------------------------ SUMMARY -------------------------------
⚠️ This filter is here to help spotting changes that should be reviewed carefully.
⚠️ It catches only index changes, deletions and value decreases.
## Deletions
 37: [-] calls: transferSpaceOwnership
43: [-] storage: pendingSpaceOwner

## Index changes
 n/a
## Decreases
 n/a
----------------------------------------------------------------------
```

**Comment by F3Joule** [comment]
Try-runtime results:

```
2024-03-14 15:23:48 initialized state externalities with storage root 0xdb0d7717d27b65bbdc8da037b58c42f631a81b006009eaf645052075f2e66a6a and state_version V0
2024-03-14 15:23:48 original spec: RuntimeString::Owned("subsocial-parachain")-40, code hash: 8a405e57cce8bd1cca762f61406c015023a3612c4b11e96a4e61afca9f5deace
2024-03-14 15:23:48 new spec: RuntimeString::Owned("subsocial-parachain")-41, code hash: e09ebd432b32fbc1fc4669975ec71c09d0073ad4aff80f8f93b03c9d863c4cec
2024-03-14 15:23:49 Running migration with current storage version StorageVersion(1) / onchain StorageVersion(0)
2024-03-14 15:23:49 Upgraded 4 records, storage to version StorageVersion(1)
2024-03-14 15:23:49 ⚠️ ParachainSystem declares internal migrations (which *might* execute). On-chain `StorageVersion(2)` vs current storage version `StorageVersion(2)`
2024-03-14 15:23:49 ⚠️ XcmpQueue declares internal migrations (which *might* execute). On-chain `StorageVersion(3)` vs current storage version `StorageVersion(2)`
2024-03-14 15:23:49 ⚠️ PolkadotXcm declares internal migrations (which *might* execute). On-chain `StorageVersion(0)` vs current storage version `StorageVersion(1)`
2024-03-14 15:23:49 ⚠️ DmpQueue declares internal migrations (which *might* execute). On-chain `StorageVersion(2)` vs current storage version `StorageVersion(1)`
2024-03-14 15:23:49 TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = (500200000000 ps, 5242880 byte), total weight = (500000000000 ps, 5242880 byte) (100.04 %, 100.00 %).
```

**Review by siman — CHANGES_REQUESTED** [review]

File: pallets/space-ownership/src/lib.rs [hunk 1]
```diff
@@ -45,14 +44,14 @@ pub mod pallet {
...
     pub enum Error<T> {
-        /// The current space owner cannot transfer ownership to themself.
+        /// The current entity owner cannot transfer ownership to themselves.
         CannotTransferToCurrentOwner,
-        /// Account is already an owner of a space.
-        AlreadyASpaceOwner,
+        /// Account is already an owner of an entity.
+        AlreadyOwner,
         /// Cannot transfer ownership, because a space is registered as an active creator.
         ActiveCreatorCannotTransferOwnership,
```
[the remark this answers is not in the document]
Comment (F3Joule) [reply pallets/space-ownership/src/lib.rs:76]: As the creator staking logic still exists on a chain, we should consider it properly. Once creator-staking is removed/refactored, we will eliminate this error case.

File: integration-tests/src/utils/space_ownership_utils.rs
[the remark this answers is not in the document]
Comment (F3Joule) [reply integration-tests/src/utils/space_ownership_utils.rs:1]: All the appropriate renames will be considered in a separate PR.

File: integration-tests/src/tests/space_ownership.rs [hunk 2]
```diff
@@ -108,7 +108,7 @@ fn accept_pending_ownership_should_fail_if_origin_is_already_an_owner() {
 
         assert_noop!(
             _accept_pending_ownership(Some(RuntimeOrigin::signed(ACCOUNT1)), None),
-            SpaceOwnershipError::<TestRuntime>::AlreadyASpaceOwner
+            SpaceOwnershipError::<TestRuntime>::NotAllowedToAcceptOwnershipTransfer,
```
[the remark this answers is not in the document]
Comment (F3Joule) [reply integration-tests/src/tests/space_ownership.rs:111]: We can do `CurrentOwnerCannotAcceptOwnershipTransfer`, then this error will be more specific.

**Review by olehmell — APPROVED** [review]

**Pull request merged**